display[all] extended usage status

undeath opened the pull request:
This PR extends the usage status shown in wallet-tool's `display[all]` method.

Previously only the status `new`, `used` and `empty` (for imported addresses) was used. This PR retrieves and shows additional status information for addresses that previously only showed `used` and have unspent outputs.

Instead of `used` this will show one of the following values for such addresses:
* `cj-out` for coinjoin outputs
* `cj-change` for coinjoin change outputs
* `deposit` for non-coinjoin outputs
* `reused` if there is more than one unspent output

To implement this some slight refactoring was done because related functionality was already implemented in the `wallet_fetch_history` function. This was extracted into the `get_tx_info` function.

note: I looked at the `joinmarket-qt` code and this did not appear to cause any issues there, but I have not tested the code with `joinmarket-qt`

undeath commented:
rebased

AdamISZ commented:
Guess it's an obvious comment, but i wonder if `used-weird` is a helpful name. perhaps `reused`, I guess you considered it?

On Qt I'd like the idea of using a different colour for reused; currently used is shown as red, maybe I could go with light red and dark red as extra warning. This is orthogonal to the PR.

Also orthogonal to the PR, but from ages ago, the one thing I really wanted to change about the display was, to show the number of confirmations including unconf.

undeath commented:
I agree `reused` is a better description for those addresses. Will change that.

Adding confirmations to the `get_tx_info` return value would be trivial, as it is included in the rpc data. But this will also require extending the WalletView (and/or related class) and Qt gui.

AdamISZ commented:
So, in testing, I generally shift a few coins from mixing depth 1 to mixing depth 4 before starting a tumbler (serves as  quick test of `-N 0` functioning correctly). This shows the change address as `deposit`, because the algo here is "if tx-is-coinjoin, (if amount is coinjoin amount, 'cj-out', else 'change'), else 'deposit'".

I don't know if it's worth worrying about that; moving coins from one mixdepth to the other without coinjoin is of course highly disadvised due to linkage implications, but thought I should mention it.

undeath commented:
Thanks for reviewing and testing (and figuring out the changes needed for jmqt)!

* fixed the bug in the `sum`/`zip` expression
* internal non-cj change is now labeled `non-cj-change`
* rebased

AdamISZ commented:
OK, tested on py2/3 and i'm happy with this; there's a missing import but easy to fix straight after.
My tests of the history function were cursory but it's not critical anyway.
